### ISSUE TITLE
Illegal application (Non-functional construction)

### ISSUE THREAD

**Issue by ANogin** [mattam82/Coq-Equations#84]
Another issue with another attempt at issue #78 workaround:
```
From Coq.Lists Require Import List.
From Equations Require Import Equations.

(* This type is from VST: https://github.com/PrincetonUniversity/VST/blob/v2.1/floyd/compact_prod_sum.v#L6 *)
Fixpoint compact_prod (T: list Type): Type :=
  match T with
  | nil => unit
  | t :: nil => t
  | t :: T0 => (t * compact_prod T0)%type
  end.

(* The rest is a nonsensical, just to give a minimalistic reproducible example *)
Inductive foo :=
| Nat : foo
| List : list foo -> foo.

Fixpoint foo_type (f:foo) : Type :=
  match f with
  | Nat => nat
  | List fs => compact_prod (map foo_type fs)
  end.

Equations num (f:foo) : forall (val:foo_type f), nat := {
  num Nat := fun val => val;
  num (List nil) := fun val => 0;
  num (List (cons hd tl)) := fun val => sum hd (num hd) tl val }
where (struct fs) sum (f:foo) (numf: (foo_type f -> nat)) (fs : list foo) (val: compact_prod (map foo_type (f::fs))) : nat := {
  sum f numf nil val := numf val;
  sum f numf (cons hd tl) val := numf (fst val) + sum hd (num hd) tl (snd val)}.
```
results in 
```
Illegal application (Non-functional construction): 
The expression "f" of type "P Nat (fun val : foo_type Nat => val)"
cannot be applied to the terms
 "f3" : "foo"
 "numf" : "foo_type f3 -> nat"
 "[]" : "list foo"
 "val" : "compact_prod (map foo_type [f3])"
 "numf val" : "nat"
```

**Comment by ANogin** [comment]
Just to add to the list of weird things that happen with these kinds of examples, if I change the above test case by swapping the order of num and sum (that is, replacing `Equations num ... where sum ...` with `Equations sum ... where num ...`), then I get
```
Recursive definition of sum is ill-formed.
In environment
sum : forall f : foo, (foo_type f -> nat) -> forall fs : list foo, compact_prod (map foo_type (f :: fs)) -> nat
f : foo
numf : foo_type f -> nat
fs : list foo
f0 : foo
l : list foo
val : compact_prod (map foo_type (f :: f0 :: l))
num : forall f : foo, foo_type f -> nat
f1 : foo
l0 : list foo
f2 : foo
l1 : list foo
val0 : foo_type (List (f2 :: l1))
Recursive call to sum has principal argument equal to "l1" instead of "l".
Recursive definition is:
"fun (f : foo) (numf : foo_type f -> nat) (fs : list foo) =>
 sum_functional sum (fix num (f0 : foo) : foo_type f0 -> nat := num_functional sum num f0) f numf fs".
```

**Comment by ANogin** [comment]
See also the "P.S." comment in #93 for an example of what makes a difference in seeing or not seeing this issue.

**Comment by ANogin** [comment]
This appears to be fixes in the latest v8.8

**Comment by mattam82** [comment]
Indeed, it's similar to #105
